### PR TITLE
Create notebook and md file in mlops-stacks to enable inference table monitoring

### PR DESCRIPTION
**Pull request by mingyu89**
python Wheel and questionnaire are not included in the code. 

## README.md

Doc
Screen Shot 2023-02-16 at 2 46 54 PM
Screen Shot 2023-02-16 at 2 47 01 PM

### Generate project 
Test generating the project. The data monitoring files exist only if include_data_monitoring is yes.
Screen Shot 2023-02-16 at 1 22 40 PM
### Follow README to copy table from metastore to catalog
Screen Shot 2023-02-16 at 2 45 27 PM

### Run notebook to set up a monitor
Only add required parameters
Screen Shot 2023-02-16 at 3 02 19 PM

Successfully run the notebook
Screen Shot 2023-02-16 at 3 04 55 PM

Monitor linked from model
Screen Shot 2023-02-16 at 3 05 58 PM

Dashboard
Screen Shot 2023-02-16 at 3 29 46 PM

Screen Shot 2023-02-16 at 3 30 25 PM

### CODE REVIEW

**Comment by mingyu89** [comment]
@vladimirk-db We got feedback that it's ok to expose the API. But not the questionnaire or python wheel. Feel free to reject the PR if you think we shouldn't add data monitoring to stack at this moment.

I also feel we should always add monitoring without asking user the question. Please let me know how you feel.

File: {{cookiecutter.project_name}}/monitoring/README.md [hunk 1]
```diff
@@ -0,0 +1,72 @@
+# Setup Data Monitoring for Inference Table
+
+**Databricks data monitoring is currently in private review. Please `sign up` for the private review and get approved
```
Comment (vperiyasamy) [inline {{cookiecutter.project_name}}/monitoring/README.md:3]:
```suggestion
**Databricks Data Monitoring is currently in Private Preview. Please `sign up` for the preview and get approved 
```

File: {{cookiecutter.project_name}}/monitoring/README.md [hunk 2]
```diff
@@ -0,0 +1,72 @@
+# Setup Data Monitoring for Inference Table
+
+**Databricks data monitoring is currently in private review. Please `sign up` for the private review and get approved 
+before proceeding.**
+
+This directory contains the notebook that can run in databricks workspace
+to set up data monitoring for prod inference table.
```
Comment (vperiyasamy) [inline {{cookiecutter.project_name}}/monitoring/README.md:7]:
```suggestion
This directory contains the notebook that can run in a Databricks workspace
to set up Data Monitoring for a production inference table.
```

File: {{cookiecutter.project_name}}/monitoring/README.md [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+## Prerequisites
+
+### Unity Catalog
+The table to be monitored must exist in Unity Catalog.
+If Unity Catalog is enabled for your workspace but the table exists in metastore, please copy the table to Unity Catalog.
```
Comment (vperiyasamy) [inline {{cookiecutter.project_name}}/monitoring/README.md:13]:
```suggestion
If Unity Catalog is enabled for your workspace but the table exists in the Hive metastore, please copy the table to Unity Catalog.
```

File: {{cookiecutter.project_name}}/monitoring/README.md [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+For example, the following code shows how to import your data to a Unity Catalog managed table
+ - load data from metastore example_project_predictions
+ - write data to Unity Catalog my_unity_catalog.my_schema.my_inference_table
+ - convert prediction column to double type
+ - convert tpep_pickup_datetime to timestamp type
+ - convert tpep_dropoff_datetime to timestamp type
```
Comment (vperiyasamy) [inline {{cookiecutter.project_name}}/monitoring/README.md:20]: Are these instructions relevant for showing how to import data into UC? Seems orthogonal

File: {{cookiecutter.project_name}}/monitoring/README.md [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+- A column that contains the predictions. For regression problems, the column must be a subclass of `NumericType`.
+- A column that contains the time of inference. This column must be of type `TimestampType`.
+- A column that contains the model version used for inference.
+
+
+## Usage
+
+### 1. Fill in data monitoring wheel URL
+The `SetUpDataMonitoringForInferenceTable.py` notebook can be found in the monitoring directory. Please complete TODO 
+and fill in the wheel_URL. The URL can be found from data monitoring private review user guide.
```
Comment (vperiyasamy) [inline {{cookiecutter.project_name}}/monitoring/README.md:45]:
```suggestion
and fill in the wheel_URL. The URL can be found from the Data Monitoring Private Preview user guide.
```

File: {{cookiecutter.project_name}}/monitoring/SetUpDataMonitoringForInferenceTable.py [hunk 6]
```diff
@@ -0,0 +1,118 @@
+# Databricks notebook source
+# MAGIC %md
+# MAGIC
+# MAGIC # Data Monitoring for Inference Table
+# MAGIC see monitoring/README.md for details
+
+# COMMAND ----------
+
+# Install data monitoring client library. Eventually this will be included in MLR
```
Comment (vperiyasamy) [inline {{cookiecutter.project_name}}/monitoring/SetUpDataMonitoringForInferenceTable.py:9]: The MLR note was for a bugbash notebook
```suggestion
# Install the Data Monitoring client library
```

File: {{cookiecutter.project_name}}/monitoring/SetUpDataMonitoringForInferenceTable.py [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+
+inference_analysis = dm.analysis.InferenceLog(
+    timestamp_col=timestamp_col,
+    model_version_col=model_version_col,
+    prediction_col=prediction_col,
+    problem_type=problem_type,
+    label_col=label_col,
+    example_id_col=example_id_col
+)
+dm.create_or_update_monitor(
```
Comment (vperiyasamy) [inline {{cookiecutter.project_name}}/monitoring/SetUpDataMonitoringForInferenceTable.py:109]: I remember we discussed in the design doc but don't remember the conclusion: do we plan to give users a separate notebook for refreshing metrics, or do we want them to schedule and run this single one? If so, I think it'll be more clear if we run this with `skip_analysis=True` and then call `refresh_metrics` directly after. This helps the user understand the behavior better, compared to the implicit refresh we do on creation.

**Review by vperiyasamy — CHANGES_REQUESTED** [review]
Overall this is looking great! Left some wording suggestions, and also have a question about refreshing metrics

File: {{cookiecutter.project_name}}/monitoring/README.md (hunk 4 above)
Comment (mingyu89) [reply {{cookiecutter.project_name}}/monitoring/README.md:20]: It describes the example, I wanted to indicate that changes can also be applied when the table is copied to UC. 
Let me remove them.

File: {{cookiecutter.project_name}}/monitoring/SetUpDataMonitoringForInferenceTable.py (hunk 7 above)
Comment (mingyu89) [reply {{cookiecutter.project_name}}/monitoring/SetUpDataMonitoringForInferenceTable.py:109]: We are depending on the instructions that users get from the private preview.

File: {{cookiecutter.project_name}}/monitoring/SetUpDataMonitoringForInferenceTable.py [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+
+# List of granularities to use when aggregating data
+# into time windows based on their timestamp. Currently the following static granularities are supported:
+# {"5 minutes", "30 minutes", "1 hour", "1 day", "n week(s)", "1 month", "1 year"}.
+granularities = ["1 hour"]
+
+# List of Databricks entity names that are associated
+# with this table. Only following entities are supported:
+#    ["models:/registry_model_name"] links a model registry model to the monitored table.
+linked_entities = ["models:/prod-{{cookiecutter.model_name}}"]
```
Comment (vperiyasamy) [inline {{cookiecutter.project_name}}/monitoring/SetUpDataMonitoringForInferenceTable.py:50]: Related to above, you're already automatically filling this in via template so should we remove from required section?

File: {{cookiecutter.project_name}}/monitoring/SetUpDataMonitoringForInferenceTable.py [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+
+# MAGIC %md
+# MAGIC 
+# MAGIC ### Required Fields
+# MAGIC TODO: Fill in and review required fields
+
+# COMMAND ----------
+
+# The name of the column that represents the inference timestamp (i.e. time of prediction) for this row. This column should be of TimestampType.
+timestamp_col = "ts"
```
Comment (vperiyasamy) [inline {{cookiecutter.project_name}}/monitoring/SetUpDataMonitoringForInferenceTable.py:23]: Since these are all required and we expect the user must fill them in, would it be better to leave them as None and then assert below that all required parameters are not none? This will avoid situations where the user doesnt study this notebook carefully and just runs it (since all required parameters are filled in) and then hits weird errors because it's trying to create a monitor with these irrelevant parameters

**Review by vperiyasamy — APPROVED** [review]
LGTM from ML Data! Left some non-blocking comments about organizing the structure of required parameters, but otherwise seems good to go. Thanks for iterating on this!

File: {{cookiecutter.project_name}}/monitoring/SetUpDataMonitoringForInferenceTable.py (hunk 9 above)
Comment (mingyu89) [reply {{cookiecutter.project_name}}/monitoring/SetUpDataMonitoringForInferenceTable.py:23]: If user uses batch inference from MLOps stacks, our BatchInference notebook adds the timestamp column with name inference_timestamp.(I'll need to update the column name here)
I can make it optional and add comments about the preset name.

Comment (vperiyasamy) [reply {{cookiecutter.project_name}}/monitoring/SetUpDataMonitoringForInferenceTable.py]: Sorry not just for the timestamp column but all required columns. The idea is that we force the user to specify them instead of providing fake values here

File: {{cookiecutter.root_dir__update_if_you_intend_to_use_monorepo}}/{{cookiecutter.project_name}}/monitoring/SetUpDataMonitoringForInferenceTable.py [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+
+# The type of ML problem that the model is solving. This should be either “classification” or “regression”.
+# Other problem types are not supported at this time. 
+problem_type = "regression"
+
+# The inference table holding data to be monitored. Table name format {catalog}.{schema}.{table}
+inference_table_name = "{catalog}.{schema}.{table}"
+
+# Name of the schema in which to create output tables. format {catalog}.{schema}
+output_schema_name = "{catalog}.{schema}"
```
Comment (vperiyasamy) [inline {{cookiecutter.root_dir__update_if_you_intend_to_use_monorepo}}/{{cookiecutter.project_name}}/monitoring/SetUpDataMonitoringForInferenceTable.py:43]: I see what you're saying about the batch inference notebook generating default column names. However, these parameters still require intervention?
- `regression`: if the user accidentally creates the monitor with regression without realizing they had to update this to classification, they'll need to manually delete and recreate
- `inference_table_name` and `output_schema_name`: these must be changed because the bracket defaults are not actual table names.

Should we make the defaults `None` and then fail the notebook if the user has not changed them?

Comment (mingyu89) [reply {{cookiecutter.root_dir__update_if_you_intend_to_use_monorepo}}/{{cookiecutter.project_name}}/monitoring/SetUpDataMonitoringForInferenceTable.py:43]: Thank you! It makes sense. I'll update it.

File: {{cookiecutter.root_dir__update_if_you_intend_to_use_monorepo}}/{{cookiecutter.project_name}}/monitoring/README.md [hunk 11]
```diff
@@ -1,3 +1,62 @@
-# Model Monitoring
+# Setup Data Monitoring for Inference Table
+
+**Databricks Data Monitoring is currently in Private Preview. Please `sign up` for the preview and get approved
```
Comment (arpitjasa-db) [inline {{cookiecutter.root_dir__update_if_you_intend_to_use_monorepo}}/{{cookiecutter.project_name}}/monitoring/README.md:3]: Any way we can provide more information or a link on how to sign up?

File: {{cookiecutter.root_dir__update_if_you_intend_to_use_monorepo}}/{{cookiecutter.project_name}}/monitoring/SetUpDataMonitoringForInferenceTable.py [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+# MAGIC %md
+# MAGIC
+# MAGIC # Data Monitoring for Inference Table
+# MAGIC see monitoring/README.md for details
+
+# COMMAND ----------
+
+# Install the Data Monitoring client library
+# TODO: Fill in wheel URL
+%pip install "{wheel_URL}"
```
Comment (arpitjasa-db) [inline {{cookiecutter.root_dir__update_if_you_intend_to_use_monorepo}}/{{cookiecutter.project_name}}/monitoring/SetUpDataMonitoringForInferenceTable.py:11]: Do we want to make this a DB widget instead to make it more obvious? Or do we not expect this notebook to be opened in Databricks?

File: cookiecutter.json [hunk 13]
```diff
@@ -10,5 +10,6 @@
   "release_branch": "Name of the release branch. The production jobs (model training, batch inference) defined in this stack pull ML code from this branch. Default: `release`",
   "read_user_group": "Display name of the user group to give read permissions to the ML jobs, integration test job runs, and machine learning resources created for this project. A group with this name must exist in both the staging and prod workspaces. Default: `users`",
   "include_feature_store": ["no", "yes"],
+  "include_data_monitoring": ["yes", "no"],
```
Comment (arpitjasa-db) [inline cookiecutter.json:13]: Do we want to flip this since by default they'd have to sign up for the private preview first?

**Review by arpitjasa-db — COMMENTED** [review]
Left a few minor comments, but overall looks great thanks @mingyu89 !

File: {{cookiecutter.root_dir__update_if_you_intend_to_use_monorepo}}/{{cookiecutter.project_name}}/monitoring/README.md (hunk 11 above)
Comment (vperiyasamy) [reply {{cookiecutter.root_dir__update_if_you_intend_to_use_monorepo}}/{{cookiecutter.project_name}}/monitoring/README.md:3]: We're discussing within our team about whether we want to expose our questionnaire to the public internet. Will update y'all when we have an answer - in the mean time, maybe we can say something like "Please reach out to your Databricks representative in order to sign up for the preview"?

File: cookiecutter.json (hunk 13 above)
Comment (mingyu89) [reply cookiecutter.json:13]: Removed the parameter.

File: {{cookiecutter.root_dir__update_if_you_intend_to_use_monorepo}}/{{cookiecutter.project_name}}/monitoring/SetUpDataMonitoringForInferenceTable.py (hunk 12 above)
Comment (mingyu89) [reply {{cookiecutter.root_dir__update_if_you_intend_to_use_monorepo}}/{{cookiecutter.project_name}}/monitoring/SetUpDataMonitoringForInferenceTable.py:11]: sure

**Review by mingyu89 — COMMENTED** [review]
Discussed with Vlad offline. 
This notebook provide benefit to users that are in both the mlops stacks private preview and the data monitoring private preview. 

We can add comment to README: "data monitoring is in private preview, please contact your databricks representative for enabling data monitoring on mlops stacks".
For users using data monitoring, they can copy the notebook from this PR.

**Comment by arpitjasa-db** [comment]
> Discussed with Vlad offline. This notebook provide benefit to users that are in both the mlops stacks private preview and the data monitoring private preview.
> 
> We can add comment to README: "data monitoring is in private preview, please contact your databricks representative for enabling data monitoring on mlops stacks". For users using data monitoring, they can copy the notebook from this PR.

@mingyu89 I see we closed this PR but we are going to have this PR as a reference? Not sure if that will be easily discoverable and looks a bit weird as well to recommend a notebook from a closed PR.
